lib/screens/v2/location_context.ex

```diff
@@ -35,100 +33,192 @@ defmodule Screens.LocationContext do
   end
 
-  defp fetch_tagged_stop_sequences_by_app(app, stop_id, _routes_at_stop)
+  defp fetch_tagged_stop_sequences_by_app(app, stop_ids, _route_ids)
        when app in [BusEink, BusShelter, GlEink] do
-    RoutePattern.fetch_tagged_stop_sequences_through_stop(stop_id)
+    fetch_tagged_stop_sequences_through_stops(stop_ids)
   end
 
-  defp fetch_tagged_stop_sequences_by_app(Dup, stop_id, routes_at_stop) do
-    route_ids = Route.route_ids(routes_at_stop)
-    RoutePattern.fetch_tagged_parent_station_sequences_through_stop(stop_id, route_ids)
+  defp fetch_tagged_stop_sequences_by_app(Dup, stop_ids, route_ids) do
+    fetch_tagged_parent_station_sequences_through_stops(stop_ids, route_ids)
   end
 
-  defp fetch_tagged_stop_sequences_by_app(PreFare, stop_id, routes_at_stop) do
-    route_ids = Route.route_ids(routes_at_stop)
-
+  defp fetch_tagged_stop_sequences_by_app(PreFare, stop_ids, route_ids) do
     # We limit results to canonical route patterns only--no stop sequences for nonstandard patterns.
-    RoutePattern.fetch_tagged_parent_station_sequences_through_stop(stop_id, route_ids, true)
+    fetch_tagged_parent_station_sequences_through_stops(stop_ids, route_ids, true)
+  end
+
+  # Returns a map from route ID to a list of stop sequences of that route, for all routes serving
+  # stop, in all applicable directions.
+  @spec fetch_tagged_stop_sequences_through_stops([Stop.id()], [Route.id()]) ::
+          {:ok, %{Route.id() => [[Stop.id()]]}} | :error
+  defp fetch_tagged_stop_sequences_through_stops(stop_ids, route_ids \\ []) do
+    case RoutePattern.fetch(%{stop_ids: stop_ids, route_ids: route_ids}) do
+      {:ok, patterns} -> {:ok, get_tagged_stop_sequences(patterns)}
+      _ -> :error
+    end
+  end
+
+  # Returns a map from route ID to a list of stop sequences of that route. Stop sequences are
+  # described in terms of parent station IDs, not platform IDs.
+  #
+  # If no parent station data exists, platform ID is returned instead. Only stop sequences for
+  # direction ID 0 are returned. Assumes that all stop sequences in result are platforms.
+  @spec fetch_tagged_parent_station_sequences_through_stops([Stop.id()], [Route.id()], boolean()) ::
+          {:ok, %{Route.id() => [[Stop.id()]]}} | :error
+  defp fetch_tagged_parent_station_sequences_through_stops(
+         stop_ids,
+         route_ids,
+         canonical? \\ false
+       ) do
+    params = %{stop_ids: stop_ids, route_ids: route_ids}
+    params = if(canonical?, do: Map.put(params, :canonical?, true), else: params)
+
+    case RoutePattern.fetch(params) do
+      {:ok, []} -> :error
+      {:ok, patterns} -> {:ok, get_tagged_parent_station_sequences(patterns)}
+      _ -> :error
+    end
+  end
+
+  # NOTE: only public due to use in tests. Should be treated as private.
+  @spec untag_stop_sequences(%{Route.id() => [[Stop.id()]]}) :: [[Stop.id()]]
+  def untag_stop_sequences(tagged_stop_sequences),
+    do: Enum.flat_map(tagged_stop_sequences, &elem(&1, 1))
```

Comment:
This use of `elem/2` on a tuple seems brittle. Especially since this is a map, should we do something like: `Enum.flat_map(tagged_stop_sequences, fn {_route_id, sequences} -> sequences end)`

Reply:
I'd agree your version makes it easier to understand what's going on. This code was copied as-is from elsewhere, but no harm in improving it.